lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -169,142 +170,234 @@ struct ProcedureOpConversion : public OpConversionPattern<ProcedureOp> {
   LogicalResult
   matchAndRewrite(ProcedureOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    Location loc = op.getLoc();
-    auto procOp = rewriter.create<llhd::ProcessOp>(loc);
-
-    // TODO: properly handle the procedure kind attribute
-    if (op.getKind() != ProcedureKind::Always)
-      return rewriter.notifyMatchFailure(loc, "not yet supported");
+    auto loc = op.getLoc();
+    if (failed(rewriter.convertRegionTypes(&op.getBody(), *typeConverter)))
+      return failure();
 
-    // Collect all event ops in the procedure.
-    SmallVector<EventOp> events(op.getOps<EventOp>());
+    // Handle initial and final procedures. These lower to a corresponding
+    // `llhd.process` or `llhd.final` op that executes the body and then halts.
+    if (op.getKind() == ProcedureKind::Initial ||
+        op.getKind() == ProcedureKind::Final) {
+      Operation *newOp;
+      if (op.getKind() == ProcedureKind::Initial)
+        newOp = rewriter.create<llhd::ProcessOp>(loc);
+      else
+        newOp = rewriter.create<llhd::FinalOp>(loc);
+      auto &body = newOp->getRegion(0);
+      rewriter.inlineRegionBefore(op.getBody(), body, body.end());
+      for (auto returnOp :
+           llvm::make_early_inc_range(body.getOps<ReturnOp>())) {
+        rewriter.setInsertionPoint(returnOp);
+        rewriter.replaceOpWithNewOp<llhd::HaltOp>(returnOp);
+      }
```

Comment:
If I remember correctly, the **return** statement cannot be used in a(n) initial/final. Slang also detects it(`return statement is only valid inside task and function blocks`).

Reply:
Yes you're right! The Moore dialect has its own `moore.return` operation, which is used to terminated `moore.procedure` ops. We should probably give it a better name… A `return` in Verilog is mapped to a `func.return` in MLIR, which is different from the `moore.return` inserted implicitly as a terminator for `moore.procedure`s 😬

Reply:
Oh! I mistook `body.getOps<ReturnOp>` as the return of function. Thanks for correcting me!